File: lab4/Lab5_wallfollowing/controllers/MyLab5Controller/MyLab5Controller.java
// MyLab5Controller.java
/*
 * MyLab5Controller Class Definition
 * File: MyLab5Controller.java
 * Date: 15th Oct 2022
 * Description: Simple Controller based on the Lab4 controller (2022)
 * Author: Terry Payne (trp@liv.ac.uk)
 */

import com.cyberbotics.webots.controller.Supervisor;
import com.cyberbotics.webots.controller.Camera;

public class MyLab5Controller {

  public static void main(String[] args) {

    Supervisor robot = new Supervisor();
    int timeStep = (int) Math.round(robot.getBasicTimeStep());
    
    Pose robot_pose = new Pose(0.0, 0.0, 0.0);
    PioneerNav2 nav = new PioneerNav2(robot, robot_pose);
    
    Camera camera = robot.getCamera("camera");
    if (camera != null)
      camera.enable(timeStep);

    PioneerProxSensors1 prox_sensors = new PioneerProxSensors1(robot, "sensor_display", robot_pose);

    double time_elapsed = 0;
    double target_time = 0;
    double robot_velocity = 0.3;
    
    // define schedule
    PioneerNav2.MoveState[] schedule = { PioneerNav2.MoveState.ARC };
    int schedule_index = -1; // we increment before selecting the current action
    PioneerNav2.MoveState state; // current state

    while (robot.step(timeStep) != -1) {
      // Testing out the front proximity sensors

      robot_pose = nav.get_real_pose();
      prox_sensors.set_pose(robot_pose);
      prox_sensors.paint();  // Render sensor Display
               
      state = nav.getState();
      if (time_elapsed > target_time) {
        time_elapsed = 0;
            
        // select next action in schedule if not stopped
        schedule_index = (schedule_index + 1) % schedule.length;
        state = schedule[schedule_index];
        
        if (state == PioneerNav2.MoveState.FORWARD) {
          target_time = nav.forward(0.8, robot_velocity);
        } else 
        if (state == PioneerNav2.MoveState.ARC) {
          target_time = nav.arc(Math.PI/2.0, 0.0, robot_velocity);
        } else
        if (state == PioneerNav2.MoveState.STOP) {
          nav.stop();
          target_time = 60 * 1000; // This doesn't really stop, but pauses for 1 minute
        }
      } else
        time_elapsed += timeStep;    // Increment by the time state
      
    };
    // Enter here exit cleanup code.
  }
}

File: lab4/controllers/my_lab5_controller/my_lab5_controller.py
"""my_lab4_controller controller."""

# my_lab4_controller Class Definition
# File: my_lab4_controller.py
# Date: 15th Nov 2022
# Description: Simple Controller based on 2021 version (2022)
# Author: Terry Payne (trp@liv.ac.uk)
#

from controller import Supervisor
import pioneer_nav2 as pn
import pioneer_proxsensors1 as pps
import math
import pose

from pioneer_nav2 import MoveState

def run_robot(robot):
        
    # get the time step of the current world.
    timestep = int(robot.getBasicTimeStep())
    
    robot_pose = pose.Pose(0.0, 0.0, 0.0)
    nav = pn.PioneerNavigation(robot, robot_pose)
    
    camera = robot.getDevice('camera')
    if camera is not None:
        camera.enable(timestep)

    prox_sensors = pps.PioneerProxSensors(robot, "sensor_display", robot_pose);

    time_elapsed = 0
    target_time = 0
    robot_velocity = 0.3

    # define schedule
    schedule = [ MoveState.ARC ]
    schedule_index = -1 # we increment before selecting the current action
     
    while robot.step(timestep) != -1:
        # Testing out the front proximity sensors

        robot_pose = nav.get_real_pose()
        prox_sensors.set_pose(robot_pose)
        prox_sensors.paint()
                    
        state = nav.state       
        if (time_elapsed > target_time):
            time_elapsed = 0
            
            # select next action in schedule if not stopped
            schedule_index = (schedule_index +1) % len(schedule)
            nav.state = schedule[schedule_index]
            
            if (nav.state == MoveState.FORWARD):
                target_time = nav.forward(0.8, robot_velocity)
            elif (nav.state == MoveState.ARC):
                target_time = nav.arc(math.pi/2.0, 0.0, robot_velocity)
            elif (nav.state == MoveState.STOP):
                nav.stop()
                target_time = 60 * 1000 # This doesn't really stop, but pauses for 1 minute
        else:
            time_elapsed += timestep    # Increment by the time state

   
if __name__ == "__main__":
    # create the Supervised Robot instance.
    my_robot = Supervisor()
    run_robot(my_robot)
